Rejection of self-referential closure signature in new solver due to lack of eager norm

```
trait Indir<T>: FnOnce(T) -> Self::Ret {
    type Ret;
}
impl<F, T, R> Indir<T> for F where F: FnOnce(T) -> R {
    type Ret = R;
}

trait Mirror {
    type Assoc<'a>;
}

fn needs<T: Mirror>(_: impl for<'a> Indir<T::Assoc<'a>>) {}

fn test<T>() where for<'a> T: Mirror<Assoc<'a> = i32> {
    needs::<T>(|x| { x.to_string(); });
}
```

there is a more general generalization issue here
```rust
trait Trait<T> {
    type Assoc<'a>;
}

impl<T> Trait<T> for () {
    type Assoc<'a> = &'a ();
}

fn foo<T>(x: Option<*mut T>) -> for<'a> fn(<() as Trait<T>>::Assoc<'a>) {
    |_| ()
}

fn main() {
    let mut x = None;
    let mut y = foo(x);
    x = Some(&mut y);
}
```
compiles with old, errors with new
```rust
error[E0308]: mismatched types
  --> src/lib.rs:18:14
   |
18 |     x = Some(&mut y);
   |              ^^^^^^ one type is more general than the other
   |
   = note: expected raw pointer `*mut fn(&())`
              found raw pointer `*mut for<'a> fn(&'a ())`

```

--- a/tests/ui/traits/next-solver/generalize/alias-with-bound-vars-incomplete-generalization.rs
+++ b/tests/ui/traits/next-solver/generalize/alias-with-bound-vars-incomplete-generalization.rs
@@ -2,7 +2,9 @@
 //@[next] compile-flags: -Znext-solver
 //@ check-pass
 
-// Regression test for the fourth variant of trait-system-refactor-initiative#191
+// Regression test for the fourth variant of trait-system-refactor-initiative#191.
+// We previously didn't normalize `<() as Trait<T>>::Assoc<'a>` before generalizing
+// here, resulting in an error.
 
 trait Trait<T> {
     type Assoc<'a>;

